### DISCUSSION TITLE
What is the workflow for using NixOS + venvs?

### DISCUSSION THREAD

**Discussion by MayNoii** [benlubas/molten-nvim#239] [Q&A] [answered]
This is a very interesting and good plugin, but I'm running into some weird problems. Add to that the complication of using NixOS. Namely:

1. Is Molten supposed to pick up on kernels installed inside the project's venv? If the answer is yes, then that isn't working for me, the only kernels it's picking up on are the ones I explicitly install using `wrapNeovimUnstable`.
2. Is the kernel supposed to pick up on Python libraries installed locally? For example, I had the `pandas` module installed in a venv, and opened up Neovim inside that venv. While my language server picked up on the module, Molten didn't, and I only could get what I wanted if I *also* installed pandas using `wrapNeovimUnstable`.

To be honest, I'm still trying to figure out how to work with NixOS + venvs in general, so apologies if these are silly questions. Thank you either way!

**Comment by benlubas** [comment, marked as answer]
Following the [venv guide](https://github.com/benlubas/molten-nvim/blob/main/docs/Virtual-Environments.md) is a good place to start. I'm also running nixOS, and I'll say that my venvs are a little weird b/c I was also confused at first.

You definitely don't need to install anything with `wrapNeovimUnstable`. You just have to install some python packages (pip in a venv, or with a nix shell both seem to work, not sure about mixing the two) and then when you run `python -m ipykernel install --user --name whatever` the kernel should "capture" all of the available packages. My understanding is also a bit shaky tbh, but that's the behavior that I've observed so far.

Your LSP and your jupyter kernel are gonna be totally separate, running different packages. This is why I'd recommend naming your kernels after the venv that they're installed from. So you can start the kernel that matches your current venv.

**Reply by MayNoii** [reply]
That seems to work, thank you! But how do you use `python -m ipykernel install --user --name whatever`? If I just run that inside the virtual environment, I get the error: 
```ImportError: libstdc++.so.6: cannot open shared object file: No such file or directory```
Which is your normal NixOS "hey, this is not a FHS distro" kind of error. I'm using poetry, if that makes any difference, but since it just uses Python venvs under the hood I don't think it should.

Anyways, I *can* work around this if I [set up an FHS environment](https://nixos-and-flakes.thiscute.world/best-practices/run-downloaded-binaries-on-nixos), but opening the FHS every time I want to use that kernel sounds really impractical.

**Reply by benlubas** [reply]
I was able to run the install command from a nix shell, and then the kernel just works. Which.. is confusing to me honestly, but I kinda just rolled with it when I got to nix.

Most of this plugins development was done on MacOS.

From what I've seen, the state of python on Nix is so bad that tossing up your hands and using an FHS environment is not uncommon
